src/miniscript/display.rs

```diff
@@ -0,0 +1,314 @@
+                },
+                Terminal::Alt(ref sub)
+                | Terminal::Swap(ref sub)
+                | Terminal::DupIf(ref sub)
+                | Terminal::Verify(ref sub)
+                | Terminal::NonZero(ref sub)
+                | Terminal::ZeroNotEqual(ref sub) => {
+                    Tree::Unary(DisplayNode::Node(sub.ty, sub.as_inner()))
+                }
+                Terminal::AndV(ref left, ref right)
+                    if matches!(right.as_inner(), Terminal::True) =>
+                {
+                    Tree::Unary(DisplayNode::Node(left.ty, left.as_inner()))
+                }
+                Terminal::OrI(ref left, ref right)
+                    if matches!(left.as_inner(), Terminal::False) =>
+                {
+                    Tree::Unary(DisplayNode::Node(right.ty, right.as_inner()))
+                }
+                Terminal::OrI(ref left, ref right)
+                    if matches!(right.as_inner(), Terminal::False) =>
+                {
+                    Tree::Unary(DisplayNode::Node(left.ty, left.as_inner()))
+                }
+                Terminal::AndV(ref left, ref right)
+                | Terminal::AndB(ref left, ref right)
+                | Terminal::OrB(ref left, ref right)
+                | Terminal::OrD(ref left, ref right)
+                | Terminal::OrC(ref left, ref right)
+                | Terminal::OrI(ref left, ref right) => Tree::Binary(
+                    DisplayNode::Node(left.ty, left.as_inner()),
+                    DisplayNode::Node(right.ty, right.as_inner()),
+                ),
+                Terminal::AndOr(ref a, ref b, ref c) if matches!(c.as_inner(), Terminal::False) => {
+                    Tree::Binary(
+                        DisplayNode::Node(a.ty, a.as_inner()),
+                        DisplayNode::Node(b.ty, b.as_inner()),
+                    )
+                }
+                Terminal::AndOr(ref a, ref b, ref c) => Tree::Ternary(
+                    DisplayNode::Node(a.ty, a.as_inner()),
+                    DisplayNode::Node(b.ty, b.as_inner()),
+                    DisplayNode::Node(c.ty, c.as_inner()),
+                ),
+                Terminal::Thresh(ref thresh) => {
+                    Tree::Nary(NaryChildren::Nodes(thresh.k(), thresh.data()))
+                }
+                Terminal::Multi(ref thresh) => {
+                    Tree::Nary(NaryChildren::Keys(thresh.k(), thresh.data()))
+                }
+                Terminal::MultiA(ref thresh) => {
+                    Tree::Nary(NaryChildren::Keys(thresh.k(), thresh.data()))
+                }
+            },
+            // Only nodes have children; the rest are terminals.
+            _ => Tree::Nullary,
+        }
+    }
+}
+
+#[derive(Copy, Clone)]
+enum DisplayTypes {
+    /// Display no types.
+    None,
+    /// Display all types, including the initial type.
+    All(Type),
```

Comment:
nit:

In cccaaec8bc189f5dcf8212c40d0f752ce90d9063, why do we need `Type` argument here? We are always calling this with `self.ty`?

Reply:
@sanket1729 yes, but the `self` of `self.ty` is a `Miniscript` while the `self` of `conditional_fmt` is a `Terminal`. So we do need to pass in `self.ty` explicitly like this. It's not available from `conditional_fmt` otherwise.

Reply:
Ah, makes sense